[WIP] feat: adding schema update methods

Shreyas220 opened the pull request:
usage
```
	us := table.UpdateSchemaBuilder(ops, base, s, lastColumnID)

	us.AddColumn( "name", 1, true, iceberg.PrimitiveTypes.Int32, "doc", 1)
	.AddColumn( "name", 1, true, iceberg.PrimitiveTypes.Int32, "doc", 1)
	.DeleteColumn("parent").UpdateColumnType("parent", iceberg.PrimitiveTypes.Int64)
	.UpdateColumnDoc("parent", "doc").UpdateColumnDefault("parent", 1)

```

laskoviymishka commented on table/update_schema.go:
```diff
@@ -191,14 +189,12 @@ func (us *UpdateSchema) Apply() *iceberg.Schema {
 }
 
-func (us *UpdateSchema) internalAddColumn(parent, name string, new_id int, required bool, dataType iceberg.Type, doc string, initialDefaultValue any) *UpdateSchema {
+func (us *UpdateSchema) InternalAddColumn(parent, name string, new_id int, required bool, dataType iceberg.Type, doc string, initialDefaultValue any) *UpdateSchema {
```
it's just `AddColumn` now, internal nature of it is artificial, no need for overcomplicated things :D 
keep it simple,

nandorKollar commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+}
+
+func (us *UpdateSchema) AddColumn(parent, name string, new_id int, required bool, dataType iceberg.Type, doc string, initialDefaultValue any) *UpdateSchema {
+	parentID := -1
```
I don't think you need parentID and full name fields here. Eventually they get a proper value in the if below now? Also, neither of them is used in the else branch.

nandorKollar commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+		}
+		_, present := us.schema.FindColumnName(parentID)
+		if present {
+			fullName = parent + "." + name
```
Full name already has the value set to this one in line 67 no? I think only the else branch could make sense here.

laskoviymishka commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+				parentField = &fields[len(fields)-1]
+			}
+		} else {
+			panic("Cannot add to non-nested DataType: " + parent + ": " + parentField.Type.String())
```
No panic please, return error instead

laskoviymishka commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+
+		for _, id := range us.deletes {
+			if id == parentID {
+				panic("Cannot add to a column that will be deleted: " + parent)
```
Same as above. Replace all panics with error.

laskoviymishka reviewed (changes requested):
Don't panic

nandorKollar commented on table/update_schema_test.go:
```diff
@@ -0,0 +1,232 @@
+package table
```
Licence header is missing from this file

nandorKollar commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+package table
```
Licence header is missing from this file.

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+	schema                   *iceberg.Schema
+	idToParent               map[int]int
+	deletes                  []int
+	updates                  map[int]*iceberg.NestedField
```
Do we need these to be pointers? Couldn't we just use `map[int]iceberg.NestedField`?

We are also missing tracking moving a column and adding a new column (additions should be tracked separately from updates)

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+	return us
+}
+
+func (us *UpdateSchema) AddColumn(parent, name string, new_id int, required bool, dataType iceberg.Type, doc string, initialDefaultValue any) *UpdateSchema {
```
should we take a potential path (`[]string`) to a column, rather than "parent" and "name"?

zeroshade replied:
we also shouldn't take the ID as an argument, we should assign new column ids internally in `UpdateSchema` instead.

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+		}
+
+		if currentField != nil {
+			foundInDeletes := false
+			for _, id := range us.deletes {
+				if id == currentField.ID {
+					foundInDeletes = true
+
+					break
+				}
+			}
+			if !foundInDeletes {
+				panic("Cannot add column, name already exists: " + parent + "." + name)
+			}
```
this doesn't make much sense to me. A user should never be doing a delete + add, they can just do an update instead. So we shouldn't have to search through the deletes here. If the column already exists, we error.

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+			Doc:      doc,
+		}
+
+		us.updates[new_id] = nestedField
```
we should do `us.adds[parentID] = append(us.adds[parentID], nestedField)`

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+
+		}
+
+		// cannot add a required column without a default value
+		if initialDefaultValue == nil && !required && !us.allowIncompatibleChanges {
+			panic("incompatible change: cannot add required column without a default value: " + name)
+		}
```
we need to validate that the initial value is of the correct type for the provided data type

nandorKollar commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+		}
+
+		if currentField != nil {
+			foundInDeletes := false
```
I think with the help of `slices` package, you can rewrite this to a more concise form:
`foundInDeletes := slices.Contains(us.deletes, currentField.ID)`

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+	return us
+}
+
+func (us *UpdateSchema) UpdateColumnType(name string, newType iceberg.Type) *UpdateSchema {
```
do we need to have separate update functions for each element? Could we just have a single `UpdateColumn` function which takes optional arguments to decide what to update?

Shreyas220 replied:
i gave it a try 
please look at the current way and let me know if thats what you were thinking ?

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+	return us
+}
+
+func (su *UpdateSchema) internalUpdateColumnRequirement(name string, isRequired bool) {
+	field := su.findForUpdate(name)
```
make sure your consistent. the other functions use `us` while this is using `su`, can we change it back to be `us`?

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+		panic("Cannot update missing column: " + name)
+	}
+
+	if (!isRequired && !field.Required) || (isRequired && field.Required) {
+		// if the change is a noop, allow it even if allowIncompatibleChanges is false
+		return
```
isn't this just `isRequired == field.Required`?

nandorKollar commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+}
+
+func (us *UpdateSchema) findField(name string) *iceberg.NestedField {
+	if us.caseSensitive {
```
nit: 

```
fieldLookupFunction := us.schema.FindFieldByNameCaseInsensitive
if us.caseSensitive {
   fieldLookupFunction = us.schema.FindFieldByName
}
 
field, ok := us.schema.FindFieldByName(fieldLookupFunction(name))
if !ok {
   return nil
}

return &field
```

nandorKollar commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+
+func (u *UpdateSchema) applyChanges() *iceberg.Schema {
+	// Helper to check if a field ID is marked for deletion
+	isDeleted := func(id int) bool {
```
No need for this function I think, this is what `slices.Contains` does.

nandorKollar commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+			return newType.Type() == "binary" // widening
+		case "fixed":
+			return newType.Type() == "binary"
+		case "time":
```
Is time -> timestamp promotion allowed? From spec, it appears to me that only date -> timestamp promotion is allowed.

Shreyas220 commented on table/update_schema.go:
```diff
@@ -0,0 +1,532 @@
+	"github.com/apache/iceberg-go"
+)
+
+type UpdateSchema struct {
```
i have added txn but need more information 
how should we handle this now ?

Shreyas220 commented:
Thanks a lot @nandorKollar @zeroshade @laskoviymishka for your review!!
really appreciate it!
I have tried to address majority of them, a couple of them are remaining 
 
summary is 
- no panics returning error now 
- moved updateschema from table to transaction 
- changes delete, add , updates to map of int (ids)
- simplified addcolumn logic 
- using 1 function for updates (@zeroshade your review here if this is what you were looking for )
- added tests

laskoviymishka commented on table/update_schema.go:
```diff
@@ -0,0 +1,426 @@
+}
+
+func (us *UpdateSchema) findForUpdate(path []string) *iceberg.NestedField {
+
```
nit: this empty spaces at the method start look odd.

laskoviymishka commented on table/update_schema.go:
```diff
@@ -0,0 +1,426 @@
+	return fmt.Errorf("default literal of type %s is not assignable to of type %s", litType.String(), prim.Type())
+}
+
+func literalFromAny(v any) (iceberg.Literal, error) {
```
this method looks like exact copy of `iceberg.NewLiteral`
why it's needed?

Shreyas220 replied:
was not able to call newLiteral directly like this here 
```
	lit := iceberg.NewLiteral(val)

```
was getting this error 

```
any does not satisfy iceberg.LiteralTypecompiler[InvalidTypeArg]
```

this was the only way i could call NewLiteral 
do let me know if there is a way 😓

laskoviymishka reviewed (commented):
Overall, LGTM, but @zeroshade and someone from core iceberg team prob needed here.

nandorKollar commented:
> Thanks a lot @nandorKollar @zeroshade @laskoviymishka for your review!! really appreciate it! I have tried to address majority of them, a couple of them are remaining
> 
> summary is
> 
> * no panics returning error now
> * moved updateschema from table to transaction
> * changes delete, add , updates to map of int (ids)
> * simplified addcolumn logic
> * using 1 function for updates (@zeroshade your review here if this is what you were looking for )
> * added tests
> 
> @zeroshade in add column or in general while accepting column name should move from `[]string` to just `string` ? and then split ` parts := strings.Split(name, ".") `
> 
> wouldnt this be easier ?
> 
> also need more a little more help in understanding how do we handle the reference of txn here ?

Thanks, I think from readability point of view, this looks much better. I need some more time to go through the details.

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,659 @@
+	return &field
+}
+
+func allowedPromotion(oldType, newType iceberg.Type) bool {
+	switch old := oldType.(type) {
+
+	case iceberg.PrimitiveType:
+		switch old.Type() {
+		case "int":
+			return newType.Type() == "long"
+		case "float":
+			return newType.Type() == "double"
+		case "decimal":
+			o := oldType.(iceberg.DecimalType)
+			n, ok := newType.(iceberg.DecimalType)
+			return ok && o.Scale() == n.Scale() && n.Precision() > o.Precision()
+		case "date":
+			return newType.Type() == "timestamp" || newType.Type() == "timestamp_ns"
+		}
+	}
+
+	return false
+}
```
iceberg-go hasn't done any updating for v3 yet. Since it's allowed in v3, but not in v2 we'd likely have to conditionally allow the promotion

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,659 @@
+			o := oldType.(iceberg.DecimalType)
+			n, ok := newType.(iceberg.DecimalType)
+			return ok && o.Scale() == n.Scale() && n.Precision() > o.Precision()
+		case "date":
```
Agreed. Let's leave this out for now and handle it separately

xixipi-lining commented:
@Shreyas220 Hi, just wondering if there's any plan to continue with this PR?

zeroshade commented on catalog/rest/rest_integration_test.go:
```diff
@@ -53,6 +53,7 @@ func (s *RestIntegrationSuite) loadCatalog(ctx context.Context) *rest.Catalog {
 		io.S3Region:          "us-east-1",
 		io.S3AccessKeyID:     "admin",
 		io.S3SecretAccessKey: "password",
+		io.S3EndpointURL:     "http://localhost:9000",
```
do we need this instead relying on the corresponding AWS_ENDPOINT env var?

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,606 @@
+	lastColumnID int
+
+	deletes map[int]struct{}
+	updates map[int]*iceberg.NestedField
+	adds    map[int][]*iceberg.NestedField
```
do these need to map to pointers instead of just mapping to values?

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,606 @@
+//
+// us.AddColumn([]string{"a","b"}, true, iceberg.StringType{}, "doc", "default")
+// us.AddColumn([]string{"a","b"}, true, iceberg.StringType{}, "doc", "default")
+func (us *UpdateSchema) AddColumn(path []string, required bool, dataType iceberg.Type, doc string, initialDefault any) (*UpdateSchema, error) {
```
We should either return just `error` or we should return just `*UpdateSchema` and defer all the actual operations to a `Build()` method. 

There's no reason to return *both* because you don't need the returned `*UpdateSchema` to continue, so we are preventing chaining while not providing anything in return. So we should embrace the Builder pattern entirely, or we should just have the single `error` return.

zeroshade reviewed (changes requested):
Few more nit picks

Shreyas220 commented:
@zeroshade 
deferred schema update operations in build

Summary of change is that the logic present in each (add, update,e, delete move ) operation is now divided into 2 logic, that are `validations `and `apply `

- `validation `will go through all the operations, collect all the errors and return 
- `apply ` will rebuild the old schema update add, update and move... etc structs 
 
and then use the same old logic to rebuild the schema and commit

xixipi-lining commented on table/update_schema.go:
```diff
@@ -0,0 +1,834 @@
+}
+
+// RemoveLastOperation removes the most recently added operation
+func (us *UpdateSchema) RemoveLastOperation() *UpdateSchema {
```
I'm a little confused — when exactly do we need to use the Reset and RemoveLastOperation methods? Exposing the internal op queues doesn't seem to offer much benefit. Maybe a simpler interface would be sufficient?

zeroshade replied:
I agree here, I don't think there's any good use for the `Reset` or `RemoveLastOperation` methods.

zeroshade commented on literals.go:
```diff
@@ -206,6 +206,37 @@ func LiteralFromBytes(typ Type, data []byte) (Literal, error) {
+		return NewLiteral(x), nil
+	case int32:
+		return NewLiteral(x), nil
+	case int64, int:
+		return NewLiteral(int64(reflect.ValueOf(x).Int())), nil
```
make these two separate cases, reflect is costly and expensive

zeroshade commented on table/table_test.go:
```diff
@@ -1420,5 +1657,4 @@ func (t *TableWritingTestSuite) TestDeleteOldMetadataNoErrorLogsOnFileFound() {
 	// validate that no error is logged
 	logOutput := logBuf.String()
 	t.NotContains(logOutput, "Warning: Failed to delete old metadata file")
-	t.NotContains(logOutput, "no such file or directory")
```
why are we removing this?

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,834 @@
+}
+
+// Operation represents a deferred schema operation
+type SchemaOperation interface {
+	Apply(builder *schemaBuilder) error
+	Validate(schema *iceberg.Schema, settings *validationSettings) error
+	String() string
+}
```
since these operate only on internal types, why are we bothering to export this type?

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,834 @@
+	op           moveOp
+}
+
+func NewUpdateSchema(txn *Transaction, s *iceberg.Schema, lastColumnID int) *UpdateSchema {
```
needs a docstring comment

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,834 @@
+	// because if validation is called multiple times, the errors will accumulate even after removing operations causing error
+	us.errors = us.errors[:0]
+
+	settings := &validationSettings{
+		allowIncompatibleChanges: us.allowIncompatibleChanges,
+		caseSensitive:            us.caseSensitive,
+	}
```
why use a pointer for this? just take `validationSettings` instead of a `*validationSettings`

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,834 @@
+		return nil, err
+	}
+
+	// No operations means no changes
+	if len(us.operations) == 0 {
+		return us.schema, nil
+	}
```
should we do this *before* calling validate?

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,834 @@
+	// for Commit Contention
+	requirements = append(requirements, AssertCurrentSchemaID(us.schema.ID))
+
+	if existingSchemaID == nil {
+		// Get the final lastColumnID from the builder
+		builder := &schemaBuilder{
+			deletes:      make(map[int]struct{}),
+			updates:      make(map[int]*iceberg.NestedField),
+			adds:         make(map[int][]*iceberg.NestedField),
+			moves:        make(map[int][]moveReq),
+			lastColumnID: us.lastColumnID,
+			baseSchema:   us.schema,
+			settings: &validationSettings{
+				allowIncompatibleChanges: us.allowIncompatibleChanges,
+				caseSensitive:            us.caseSensitive,
+			},
+		}
+
+		for _, op := range us.operations {
+			if err := op.Apply(builder); err != nil {
+				return nil, nil, fmt.Errorf("failed to recompute lastColumnID: %w", err)
+			}
+		}
+		updates = append(updates,
+			NewAddSchemaUpdate(newSchema, builder.lastColumnID, false),
+			NewSetCurrentSchemaUpdate(newSchema.ID),
+		)
```
This seems like a waste to have to *rebuild* the schema again. Can we somehow store/update the `lastColumnID` after calling `Build()` rather than redo the build here.

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,834 @@
+		}
+	}
+
+	if len(us.errors) > 0 {
+		return fmt.Errorf("validation failed with %d errors: %v", len(us.errors), us.errors)
+	}
+	return nil
```
`return errors.Join(us.errors)` much more concise and simple :smile:

zeroshade commented on table/update_schema.go:
```diff
@@ -0,0 +1,834 @@
+}
+
+func (us *UpdateSchema) getNameMappingUpdates(newSchema *iceberg.Schema) []Update {
+	// Only update name mapping if we have adds/updates that might need it
+	hasAdds := false
+	hasUpdates := false
```
we're not doing different behaviors based on `hasAdds` vs `hasUpdates` so just use one variable

batyrchary commented:
@Shreyas220  Hi, I was wondering if you are still working on this PR ?
